Migration of sale_line_quantity_properties_based

tafaRU opened the pull request:
This PR depends on https://github.com/OCA/sale-workflow/pull/147 which also depends on https://github.com/OCA/sale-workflow/pull/167 (both merged)

tafaRU committed:
Add code using by 'git checkout origin/7.0 -- sale_line_quantity_properties_based

tafaRU committed:
Fix field name, changed in V8

tafaRU committed:
Inherit view from sale_mrp instead of sale_stock

tafaRU committed:
Update copyright year

tafaRU committed:
Extract the description in README.rst

tafaRU committed:
Add OCA icon

tafaRU committed:
Rem sale_mrp from depends because sale_properties_easy_creation already depends by it

eLBati committed:
[FIX] using new API for property_ids onchange

tafaRU committed:
Fix onchange method name and remove brackets from exec

gurneyalex commented on sale_line_quantity_properties_based/sale_order_line.py:
```diff
@@ -0,0 +1,82 @@
+                }
+                try:
+                    exec self.product_id.quantity_formula_id.\
+                        formula_text in localdict
```
use `safe_eval` (from `openerp.tools.safe_eval`) in `exec` mode. Not perfect, but limits the code injection risks.

tafaRU replied:
@gurneyalex, done at https://github.com/tafaRU/sale-workflow/commit/bfe49055d0e63e37041bf99ecfb36a5584c263c6
Sorry to bother you but https://github.com/OCA/sale-workflow/pull/167 needs to be merged asap so that I can unblock this one, https://github.com/OCA/sale-workflow/pull/147 and https://github.com/OCA/sale-workflow/pull/145

Thanks!

tafaRU committed:
Add product_id_change to avoid product_uos_qty recomputation in the same way done for v7.0

tafaRU committed:
Fix wrong interoperability among onchange (v8) and on_change (v7).
To do this I apply the following changes:
- depends from sale_properties_dynamic_fields instead of sale_properties_easy_creation
- onchange (v8) only depends by property_ids
- get empty properties and get empty properties dynamic fields in on_change (v7)

tafaRU committed:
Delete sale_order_view.xml since is useless now

tafaRU committed:
Add product_uos_qty as parameter of onchange to recompute quantity

tafaRU committed:
Quantity recalculation has to be done when property_ids or product_uos_qty

tafaRU committed:
Calculate quantity by using compute_formula method

tafaRU committed:
Improve README and update module version according to last OCA conventions

gurneyalex commented on sale_line_quantity_properties_based/README.rst:
```diff
@@ -0,0 +1,89 @@
+    10 [pcs of] (4 m x 0.5 m) shelves = 20 m² of wood
+
+In order to have this function working, it is necessary to have the user
+proceeding as follows:
```
I think s/proceeding/proceed/ 

(but I'm French...)

gurneyalex commented on sale_line_quantity_properties_based/README.rst:
```diff
@@ -0,0 +1,89 @@
+(Note: this can be more easily achieved by using the modules
+'sale_properties_easy_creation' and/or 'sale_properties_dynamic_fields')
+
+Properties must respond the following criteria:
```
s/respond/respect/

gurneyalex commented on sale_line_quantity_properties_based/sale_order_line.py:
```diff
@@ -0,0 +1,103 @@
+                    self._cr, self._uid, 'Product UoS')
+                for prop in self.env['mrp.property'].with_context(
+                    prop_ctx
+                ).browse([p.id for p in self.property_ids]):
```
[p.id for p in self.property_ids] -> self.property_ids.ids

gurneyalex commented on sale_line_quantity_properties_based/sale_order_line.py:
```diff
@@ -0,0 +1,103 @@
+                        amount, precision_digits=uom_precision)
+                except ValueError as e:
+                    _logger.warning(
+                        "Formula evaluation error: '%s'" % e.message)
```
I think this onchange method should return a `{'warning': {'title': 'formula evaluation error', 'message': more details...}}` whe ValueError is caught, so that the user gets some feedback, rather than silently doing nothing.

tafaRU committed:
Modify .travis to test sale__line_quantity_properties_based separately

tafaRU committed:
Fix typos in README file

tafaRU committed:
Improve quantity_property_ids_changed: use .ids instead of comprehension list

tafaRU committed:
Return warning instead of add it to the log, so that the user gets some feedback, rather than silently doing nothing.

tafaRU committed:
Update it.po

tafaRU commented:
@gurneyalex, I addressed your last remarks and the PR now should be ready to be merged.
Please have a look at it.

eLBati commented on .travis.yml:
```diff
@@ -35,6 +35,8 @@ env:
   - TESTS="1" ODOO_REPO="odoo/odoo" INCLUDE="sale_properties_dynamic_fields" LINT_CHECK="0"
   - TESTS="1" ODOO_REPO="OCA/OCB" INCLUDE="sale_properties_dynamic_fields" LINT_CHECK="0"
+  - TESTS="1" ODOO_REPO="odoo/odoo" INCLUDE="sale__line_quantity_properties_based" LINT_CHECK="0"
+  - TESTS="1" ODOO_REPO="OCA/OCB" INCLUDE="sale__line_quantity_properties_based" LINT_CHECK="0"
```
too many underscores

eLBati replied:
Also, I think `sale_properties_dynamic_fields` and `sale_line_quantity_properties_based` can be tested together if possible

tafaRU commented:
@eLBati, done, thanks!

eLBati commented:
@tafaRU please check tests

eLBati commented:
Maybe a rebase can help, after https://github.com/OCA/sale-workflow/pull/266

tafaRU commented:
@eLBati, just rebased, let's see.

tafaRU commented:
@eLBati travis is still red, further investigation is needed then.

tafaRU committed:
Add uos_id to localdict

atchuthan commented:
@tafaRU In .travis.yml, why do we still test with INCLUDE/EXCLUDE?
https://github.com/OCA/sale-workflow/pull/279/commits/d6b210afb96d40fdd7a17351e4d08111cfabf3ea

@pedrobaeza suggested on removing INCLUDE/EXCLUDE as it was designed for older testing versions in OCA and to adapt our travis for new version test.

tafaRU commented:
@atchuthan, thanks for your help

Could you please give me some reference about this?

> @pedrobaeza suggested on removing INCLUDE/EXCLUDE as it was designed for older testing versions in OCA and to adapt our travis for new version test.

tafaRU committed:
[FIX] KeyError: 'product_uom_qty'

gurneyalex commented:
:+1:

pedrobaeza commented:
You finally made it!

I will prefer as you know some commit squashing and module renaming, but after all the work, I approve this.
